src/core/net/ip6_mpl.cpp

```diff
@@ -334,6 +334,27 @@ uint8_t Mpl::DetermineMaxRetransmissions(void) const
     return maxRetx;
 }
 
+bool Mpl::RemoveMatchedMessage(Message &aMessage)
+{
+    bool     match = false;
+    uint16_t len   = aMessage.GetLength();
+    VerifyOrExit(len > 0);
+    for (Message &message : mBufferedMessageSet)
+    {
+        if (message.GetLength() - sizeof(Metadata) >= len)
+        {
+            if (aMessage.CompareBytes(0, message, message.GetLength() - sizeof(Metadata) - len, len))
```

Comment:
The `aMessage` we get here is already gone through IPv6 and UDP layers and all those header are stripped/removed from it. It contains the UDP payload.

I see that you are accounting for this and compare byte in `message` accordingly from offset `message.GetLength() - sizeof(Metadata) - len` (practically checking if bytes at the end of `message` match bytes in `aMessage`).  

I wonder if this is always safe/correct to do. Can we have an unintended match here? 

If we could be sure that both message are TMF (using TMF port), I feel it may be safer then to compare the UDP payloads. Thoughts?

Reply:
We are comparing around 30 bytes, for the entire TMF payload, do you think it can still have a risk of false match? I was trying to keep it generic and avoid any specific flags passed to the MPL module, but sure we can add that if you think it is unsafe. 

The approach I was thinking of, is to pass a boolean to indicate it is a TMF message in the method _RemoveMplEntry_ and then use than that to compare the port information in MPL module ( _Mpl::RemoveMatchedMessage_)to check if it matches TMF port.  The challenge here is, we wouldn't have parsed the UDP header yet while adding the MPL message to the buffered set (in _HandleExtensionHeaders_-> Ip6::HandleOptions) and may need to parse it now, and do that again while _HandleDatagram_ which becomes a bit messy?